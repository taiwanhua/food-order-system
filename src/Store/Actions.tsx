import { AnyAction } from "redux";
import { tuple } from "../Type/type";

export interface AnyActions<T, S> extends AnyAction {
    [key: string]: any;
    type: T;
    payload: S;
}

const IglobalsAction = tuple('GLOBAL1', 'GLOBAL12');

export type globalsAction = typeof IglobalsAction[number];

const ICartAction = tuple('CART_ADD', 'CART_MINUS', 'CART_CLEAR', 'CART_SUBMIT');

export type cartsAction = typeof ICartAction[number];

const IHistoryAction = tuple('HISTORY_CLEAR', 'CART_SUBMIT');

export type historyAction = typeof IHistoryAction[number];
