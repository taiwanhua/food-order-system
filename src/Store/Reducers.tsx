import { globalsAction, cartsAction, historyAction, AnyActions } from "./Actions";


export const globals = (state = initialGlobals, action: AnyActions<globalsAction, IinitialGlobalsPayload>) => {
    switch (action.type) {
        default:
            return state;
    }
}

const initialGlobals: IinitialGlobals = {

}

export interface IinitialGlobals {

}

export interface IinitialGlobalsPayload {

}

export const carts = (state = initialCarts, action: AnyActions<cartsAction, IinitialCartsPayload>) => {
    switch (action.type) {
        case 'CART_ADD':
            return {
                ...state,
                [action.payload.name]: (state?.[action.payload.name] ?? 0) + 1
            };
        case 'CART_MINUS':
            if (state?.[action.payload.name] === 1) {
                delete state[action.payload.name];
                return {
                    ...state
                }
            }
            else {
                return {
                    ...state,
                    [action.payload.name]: state[action.payload.name] - 1
                }
            }
        case 'CART_CLEAR':
            return {}
        case 'CART_SUBMIT':
            return {}
        default:
            return state;
    }
}

const initialCarts: IinitialCarts = {

}


export interface IinitialCarts {
    [K: string]: number;
}

export interface IinitialCartsPayload {
    /** 商品名稱 */
    name: string;
    /** 數量 */
    count?: number;
}

export const history = (state = initialhistory, action: AnyActions<historyAction, IinitialhistoryPayload>) => {
    switch (action.type) {
        case 'HISTORY_CLEAR':
            return {
                count: 0,
                order: {}
            };
        case 'CART_SUBMIT':
            return {
                ...state,
                count: state.count + 1,
                order: {
                    ...state.order,
                    [`#${state.count + 1}`]: { ...action.payload.order }
                }
            }
        default:
            return state;
    }
}

const initialhistory: Iinitialhistory = {
    count: 0,
    order: {}
}

export interface Iinitialhistory {
    count: number;
    order: {
        [K: string]: TOrder
    };
}

export type TOrder = {
    [K: string]: number;
}

export interface IinitialhistoryPayload {
    order: TOrder;
    count: number;
}
