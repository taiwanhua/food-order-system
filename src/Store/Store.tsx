import * as React from 'react';
import {
    Provider,
} from 'react-redux'
import { createStore, combineReducers } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension';
import { globals, carts, history } from './Reducers'

const combinedReducer = combineReducers({
    globals,
    carts,
    history
})

export type RootState = ReturnType<typeof combinedReducer>

const Stroe = createStore(combinedReducer, composeWithDevTools());

interface IProviderProps {
    /**
     * 需要使用本Context的所有子組件
     */
    children?: React.ReactNode;
}

export const ReduxProvider: React.FC<IProviderProps> = (props) => {

    return (
        <Provider store={Stroe}>
            {props.children}
        </Provider>
    )
}