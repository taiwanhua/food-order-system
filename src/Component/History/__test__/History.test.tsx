import * as React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import { History } from '..';
import { ReduxProvider } from '../../../Store/Store';

it('render History without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(<ReduxProvider>
        <History close={() => { }} />
    </ReduxProvider>, div);
});

it('match History Snapshot', () => {

    const tree = renderer.create(
        <ReduxProvider>
            <History close={() => { }} />
        </ReduxProvider >
    ).toJSON();

    expect(tree).toMatchSnapshot();

});
