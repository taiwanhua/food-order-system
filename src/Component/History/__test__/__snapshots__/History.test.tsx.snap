// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match History Snapshot 1`] = `
Array [
  <div
    className="history-bg"
    data-testid="history-bg"
    style={
      Object {
        "height": "768px",
      }
    }
  />,
  <div
    className="history-content"
  >
    <p>
      No history ! please place an order firstÔºÅ
    </p>
    <button
      className="history-close"
      data-testid="history-close"
      onClick={[Function]}
    >
      close
    </button>
    <button
      className="history-clear"
      data-testid="history-clear"
      onClick={[Function]}
    >
      clear
    </button>
  </div>,
]
`;
